pallets/collator-selection/src/lib.rs

```diff
@@ -331,14 +331,35 @@ pub mod pallet {
 			T::PotId::get().into_account()
 		}
 
+		pub fn try_remove_candidate(who: &T::AccountId) -> Result<usize, DispatchError> {
+			let current_count = <Candidates<T>>::try_mutate(|candidates| -> Result<usize, DispatchError> {
+				let index = candidates.iter().position(|candidate| candidate.who == *who).ok_or(Error::<T>::NotCandidate)?;
+				T::Currency::unreserve(&who, candidates[index].deposit);
+				candidates.remove(index);
+				Ok(candidates.len())
+			});
+			Self::deposit_event(Event::CandidateRemoved(who.clone()));
+			current_count
+		}
+
 		/// Assemble the current set of candidates and invulnerables into the next collator set.
 		///
 		/// This is done on the fly, as frequent as we are told to do so, as the session manager.
 		pub fn assemble_collators() -> Vec<T::AccountId> {
 			let mut collators = Self::invulnerables();
+			let boot_block = Self::boot_block();
 			collators.extend(
-				Self::candidates().into_iter().map(|c| c.who).collect::<Vec<_>>(),
+				Self::candidates().into_iter().filter_map(|c| {
+					if c.last_block > boot_block {
+						Some(c.who)
+					} else {
+						let _ = Self::try_remove_candidate(&c.who);
```

Comment:
```suggestion
						let outcome = Self::try_remove_candidate(&c.who);
						if let Err(why) = outcome { 
							log::warn!("..."); 
							debug_assert!(false, "expanation...");
						}
```

Reply:
or if you feel confident enough that `try_remove_candidate` can never fail, you can `.expect`, but I don't recommend this.